### DISCUSSION TITLE
How to use FileSavePicker with C++/WinRT for desktop app?

### DISCUSSION THREAD

**Discussion by ZhouJJJKKK** [microsoft/WindowsAppSDK#1434] [Q&A] [answered]
I want to open a dialog and save files with C++/WinRT for desktop app, here is my code:
```cpp
Windows::Foundation::IAsyncAction click_button(Windows::Foundation::IInspectable const& sender, winrt::Microsoft::UI::Xaml::RoutedEventArgs const& e) {
    Windows::Storage::Pickers::FileSavePicker savepicker;
    savepicker.SuggestedStartLocation(Windows::Storage::Pickers::PickerLocationId::DocumentsLibrary);
    auto plainTextExtensions = winrt::single_threaded_vector<winrt::hstring>();
    plainTextExtensions.Append(winrt::param::hstring(L".jpg"));
    savepicker.FileTypeChoices().Insert(L"jpg", plainTextExtensions);
    savepicker.SuggestedFileName(L"new ducument");
    co_await winrt::resume_background();
    auto file = co_await savepicker.PickSaveFileAsync();
    co_await resume_foreground(DispatcherQueue());
}
```
But no dialog showed, and no error occurred.
I guess the `savepicker.PickSaveFileAsync()` function was blocked, how to resolve it?

**Comment by andrewleader** [comment, marked as answer]
Hi @ZhouJJJKKK, sorry for the delayed response. In desktop-based apps, you have to use InitializeWithWindow on pickers like FilePicker to let the picker know which window (hwnd) it should display over.

```cpp
    auto picker = winrt::Windows::Storage::Pickers::FileOpenPicker();
    picker.as<IInitializeWithWindow>()->Initialize(hwnd);
    ...
```

Here's some more info: https://docs.microsoft.com/windows/apps/desktop/modernize/desktop-to-uwp-supported-api#classes-that-use-iinitializewithwindow

**Reply by nickrandolph** [reply]
I'm pretty certain this isn't the first time this feedback has been given - is it really that hard to provide a clean abstraction where developers aren't responsible for cobbling together WinRT pickers/dialogs with the hwnd of the window. This is a) super ugly to have to do every time and b) not intuitive to developers trying to use this for the first time

**Reply by andrewleader** [reply]
@nickrandolph you're correct, we've got this feedback tracked in our internal productboard (which we're hopefully sharing our public portal soon, it helps us aggregate feedback across a variety of sources). We logged your previous feedback from https://github.com/microsoft/WindowsAppSDK/issues/466#issuecomment-882687122 in our item in productboard.

We're heads-down focused on finishing 1.0 right now, this is definitely something we'd like to improve in the future.

**Reply by ZhouJJJKKK** [reply]
thank you @andrewleader  for the guide. The picker's UI successfully showed in my desktop program.

**Reply by benschop-it** [reply]
I still get a crash calling FileSavePicker.PickSaveFileAsync(). VS2022, 10.0.20348.0, CPPWinRT 2.0.220131.2, despite using InitializeWithWindow. I am testing with this project: https://github.com/robmikh/Win32CaptureSample.git. It crashes when you clikc on "Take Snapshot". I've upgraded to the versions mentioned above but it made no difference. The exception I get is Invalid window handle.
